pep-9999.rst

```diff
@@ -0,0 +1,342 @@
+PEP: 9999
+Title: Lazy Imports
+Author: Germán Méndez Bravo <german.mb@gmail.com>, Carl Meyer <carl.j.meyer@gmail.com>
+Sponsor: Barry Warsaw <barry@python.org>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 29-Apr-2022
+Python-Version: 3.12
+Discussions-To: https://discuss.python.org
+
+Abstract
+========
+
+This PEP proposes an opt-in experimental feature to transparently defer the
+execution of imported modules until the moment when an imported object is used.
+Since Python programs commonly import many more modules than a single
+invocation of the program is likely to use in practice, lazy imports can
+greatly reduce the overall number of modules loaded, improving startup time and
+memory usage. Lazy imports also mostly eliminate the risk of import cycles.
+
+
+Motivation
+==========
+
+Common Python code style prefers imports at module level, so they don't have
+to be repeated within each scope the imported object is used in, and to avoid
+the inefficiency of repeated execution of the import opcodes at runtime. This
+means that importing the main module of a program typically results in an
+immediate cascade of imports of most or all of the modules that may ever be
+needed by the program.
+
+Consider the example of a Python command line program with a number of
+subcommands. Each subcommand may perform different tasks, requiring the import
+of different dependencies. But a given invocation of the program will only
+execute a single subcommand, or possibly none (i.e. if just ``--help`` usage
+info is requested.) Top-level eager imports in such a program will result in
+the import of many modules that will never be used at all; the time spent
+(possibly compiling and) executing these modules is pure waste.
+
+Some large Python CLIs, in an effort to improve startup time, make some imports
+lazy by manually placing imports inline into functions to delay imports of
+expensive subsystems. This manual approach is labor-intensive and fragile; one
+misplaced import or refactor can easily undo painstaking optimization work.
+
+Existing import-hook-based solutions such as demandimport [1] are limited in
+that only certain styles of import can be made truly lazy (imports such as
+``from foo import a, b`` will still eagerly import the module ``foo``) and they
+impose additional runtime overhead on every module attribute access.
+
+This PEP proposes a more comprehensive solution for lazy imports that does not
+impose detectable overhead in real-world use. The implementation in this PEP
+has already demonstrated startup time wins up to 70% and memory-use wins up to
+40% on real-world Python CLIs.
```

Comment:
Is there a citation for this?

Reply:
Does the Cinder project have any published metrics that you can refer to?